src/test/java/com/google/firebase/ThreadManagerTest.java

```diff
@@ -0,0 +1,227 @@
+  }
+
+  @Test
+  public void testAppLifecycleWithServiceCall() {
+    MockThreadManager threadManager = new MockThreadManager();
+
+    // Initializing an app doesn't initialize any threading resources.
+    FirebaseApp app = FirebaseApp.initializeApp(buildOptions(threadManager));
+    assertEquals(0, threadManager.events.size());
+
+    // Initializing auth should not initialize threading resources.
+    FirebaseAuth auth = FirebaseAuth.getInstance(app);
+    assertEquals(0, threadManager.events.size());
+
+    // Calling an async method on auth should initialize threading resources.
+    try {
+      auth.verifyIdTokenAsync("foo").get();
+    } catch (InterruptedException | ExecutionException ignored) {
+      // ignored
+    }
+    assertEquals(1, threadManager.events.size());
+    Event event = threadManager.events.get(0);
+    assertEquals(Event.TYPE_GET_EXECUTOR, event.type);
+    assertEquals(app, event.arguments[0]);
+
+    // Deleting app should tear down threading resources.
+    app.delete();
+    assertEquals(2, threadManager.events.size());
+    event = threadManager.events.get(1);
+    assertEquals(Event.TYPE_RELEASE_EXECUTOR, event.type);
+    assertEquals(app, event.arguments[0]);
+  }
+
+  @Test
+  public void testAppLifecycleWithMultipleServiceCalls()
+      throws ExecutionException, InterruptedException {
+    MockThreadManager threadManager = new MockThreadManager();
+    FirebaseApp app = FirebaseApp.initializeApp(buildOptions(threadManager));
+
+    // Initializing an app doesn't initialize any threading resources.
+    assertEquals(0, threadManager.events.size());
+
+    // Initializing auth should not initialize threading resources.
+    FirebaseAuth auth = FirebaseAuth.getInstance(app);
+    assertEquals(0, threadManager.events.size());
+
+    // Calling an async method on auth should initialize threading resources.
+    try {
+      auth.verifyIdTokenAsync("foo").get();
+    } catch (InterruptedException | ExecutionException ignored) {
+      // ignored
+    }
+    assertEquals(1, threadManager.events.size());
+    Event event = threadManager.events.get(0);
+    assertEquals(Event.TYPE_GET_EXECUTOR, event.type);
+    assertEquals(app, event.arguments[0]);
+
+    // Calling a method again should not re-initialize threading resources.
+    try {
+      auth.verifyIdTokenAsync("foo").get();
+    } catch (InterruptedException | ExecutionException ignored) {
+      // ignored
+    }
+    assertEquals(1, threadManager.events.size());
+
+    // Should not re-initialize threading resources.
+    TestOnlyImplFirebaseTrampolines.forceThreadManagerInit(app);
+    assertEquals(1, threadManager.events.size());
+
+    // Deleting app should tear down threading resources.
+    app.delete();
+    assertEquals(2, threadManager.events.size());
+    event = threadManager.events.get(1);
+    assertEquals(Event.TYPE_RELEASE_EXECUTOR, event.type);
+    assertEquals(app, event.arguments[0]);
+  }
+
```

Comment:
Can you add tests that exercise the "setup-teardown-setup-teardown" logic you implemented?

Reply:
I think that's covered in the `FirebaseExecutorsTest`. Let me know if that's not what you were referring to.